<div style="border:solid purple 2px; padding: 20px">

Виктория привет! 👋 Поздравляю с первым выполненным проектом.
    
Меня зовут Рустам Муртазин, и я буду делать ревью проектной работы по предобработке данных. Давай будем общаться на **«ты»**. Если это неприемлемо, обязательно напиши мне в комментариях — мы перейдем на **«вы»**.

Ты проделала большую работу над проектом, однако в любом деле всегда можно можно что-то улучшить. Моя основная цель — не показать, что где-то совершены какие-либо ошибки, а поделиться своим опытом, который поможет тебе в дальнейших работах. Далее в файле ты сможешь увидить мои комментарии, постарайся учесть их при выполнении следующих заданий.    

Обрати внимание в первую очередь на те, что помечаны <span style="color:red">красным цветом</span>. После их доработки проект будет принят. <span style="color:green">Зеленым цветом</span> отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. <span style="color:orange">Оранжевым цветом</span> выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: если ты **что-то меняешь** в проекте по моим рекомендациям — **пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее».

В целом работа хорошая, видно, что ты старалась. Дорабатывай комментарии и отправляй работу на повторное ревью, я буду ждать 😉

</div>

<div style="border:solid PaleVioletRed 2px; padding: 20px">

<span style='color:black'> Привет, Рустам! Приятно познакомиться. Спасибо за быструю проверку, комментарии и подбадривающий тон!
Я выбрала розовый цвет для своих ответов (никаких гендерных стереотипов)!  </span>
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 </h2>

Спасибо, за быструю и качественную доработку комментариев! Все моменты были исправлены, так что работу я принимаю. Ты большая молодец, что исправила все комментарии (и оранжевые тоже). Приятно быть полезным)
    
В общем получилась классная работа. Благодарю за старания и интерес и желаю успехов в дальнейших проектах 😊

</div>

## Исследование надёжности заёмщиков

Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели **кредитного скоринга** — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.

### Шаг 1. Откройте файл с данными и изучите общую информацию. 

In [1]:
import pandas as pd #Прочитаем файл data.csv и сохраним его в переменной df.
df = pd.read_csv('/datasets/data.csv')

In [2]:
df.info()  #Получим общую информацию о данных таблицы и выведем на экран первых 10 строк таблицы.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
children            21525 non-null int64
days_employed       19351 non-null float64
dob_years           21525 non-null int64
education           21525 non-null object
education_id        21525 non-null int64
family_status       21525 non-null object
family_status_id    21525 non-null int64
gender              21525 non-null object
income_type         21525 non-null object
debt                21525 non-null int64
total_income        19351 non-null float64
purpose             21525 non-null object
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


In [3]:
display(df.head(10))  

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Совет на будущее: не увлекайся массовым выводом через функцию `print`. Для таблиц лучше использовать встроенные возможности jupyter notebook (просто вызывая имя переменной с датасетом), либо использовать функцию `display`. Например так:
    
`df.head()`
    
`display(df.info())`

</div>

<div style="border:solid PaleVioletRed 2px; padding: 20px"> Отображение результатов с помощью функции display действительно удобнее. Спасибо!  </div>

Всего в таблице 12 столбцов (признаков) и 21525 рядов (данных о клиентах). 

Разберём, какие в df столбцы и какую информацию они содержат:

- children — количество детей в семье - количественная переменная - целое число
- days_employed — общий трудовой стаж в днях - количественная переменная - число с плавающей точкой
- dob_years — возраст клиента в годах - количественная переменная - целое число
- education — уровень образования клиента -категориальная переменная -  объект
- education_id — идентификатор уровня образования - целое число
- family_status — семейное положение  - категориальная переменная - объект
- family_status_id — идентификатор семейного положения  - категориальная переменная -  целое число
- gender — пол клиента  -категориальная переменная - объект
- income_type — тип занятости  -категориальная переменная -  объект
- debt — имел ли задолженность по возврату кредитов - категориальная переменнная -целое число
- total_income — ежемесячный доход -  количественная переменная - число с плавающей точкой
- purpose — цель получения кредита  - категориальная переменная - объект

Почти все столбцы не содержат пропусков, за исключением двух days_employed и total_income. 

Рассмотрим количественные переменные более подробно.

In [4]:
df.describe()

children  days_employed     dob_years  education_id  \
count  21525.000000   19351.000000  21525.000000  21525.000000   
mean       0.538908   63046.497661     43.293380      0.817236   
std        1.381587  140827.311974     12.574584      0.548138   
min       -1.000000  -18388.949901      0.000000      0.000000   
25%        0.000000   -2747.423625     33.000000      1.000000   
50%        0.000000   -1203.369529     42.000000      1.000000   
75%        1.000000    -291.095954     53.000000      1.000000   
max       20.000000  401755.400475     75.000000      4.000000   

       family_status_id          debt  total_income  
count      21525.000000  21525.000000  1.935100e+04  
mean           0.972544      0.080883  1.674223e+05  
std            1.420324      0.272661  1.029716e+05  
min            0.000000      0.000000  2.066726e+04  
25%            0.000000      0.000000  1.030532e+05  
50%            0.000000      0.000000  1.450179e+05  
75%            1.000000      0.000000  2.034351e+05  
max            4.000000      1.000000  2.265604e+06

Минимальное значение для столбцов children и days_employed отрицательное, скорее всего это ошибка.

Теперь посмотрим на категориальные переменные. 

In [38]:
def check_unique(data):
    print(data.name, ":",  data.unique())  #метод unique позволит нам увидеть список всех вариантов для переменной

cat_var = [df['education'],df['children'],df['gender'],df['family_status'],df['income_type'],df['purpose'],df['debt']]  


for var in cat_var:
    display(check_unique(var))


education : ['высшее' 'среднее' 'неоконченное высшее' 'начальное' 'ученая степень']


None

children : [1 0 3 2 4 5]


None

gender : ['F' 'M' 'XNA']


None

family_status : ['женат / замужем' 'гражданский брак' 'вдовец / вдова' 'в разводе'
 'Не женат / не замужем']


None

income_type : ['сотрудник' 'пенсионер' 'компаньон' 'госслужащий' 'безработный'
 'предприниматель' 'студент' 'в декрете']


None

purpose : ['покупка жилья' 'приобретение автомобиля' 'дополнительное образование'
 'сыграть свадьбу' 'операции с жильем' 'образование'
 'на проведение свадьбы' 'покупка жилья для семьи' 'покупка недвижимости'
 'покупка коммерческой недвижимости' 'покупка жилой недвижимости'
 'строительство собственной недвижимости' 'недвижимость'
 'строительство недвижимости' 'на покупку подержанного автомобиля'
 'на покупку своего автомобиля' 'операции с коммерческой недвижимостью'
 'строительство жилой недвижимости' 'жилье'
 'операции со своей недвижимостью' 'автомобили' 'заняться образованием'
 'сделка с подержанным автомобилем' 'получение образования' 'автомобиль'
 'свадьба' 'получение дополнительного образования' 'покупка своего жилья'
 'операции с недвижимостью' 'получение высшего образования'
 'свой автомобиль' 'сделка с автомобилем' 'профильное образование'
 'высшее образование' 'покупка жилья для сдачи' 'на покупку автомобиля'
 'ремонт жилью' 'заняться высшим образованием']


None

debt : [0 1]


None

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Здесь ты повторяешь одно и то же действие несколько раз, возможно тогда стоит использовать для этого циклы, либо написать для этого отдельную функцию 🤔

</div>

<div style="border:solid PaleVioletRed 2px; padding: 20px">
Добавила и функцию, и цикл.
Появилась строка с None. Откуда она я не знаю. Убрать не получается. 

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 </h2>

Отличная реализация 👍 None вылезает, потому что ты использаушь комбинацию `display` и `print`. Можно убрать `display`, а для того, чтобы вывод разделялся, можно в принт добавить `\n`. Вот так `print(data.name, ":",  data.unique(), '\n')`

</div>

### Вывод

Каждая строка таблицы содержит информацию о заемщиках банка. Наша задача выяснить, как семейный статус и наличие детей влияет на платежеспособность заемщика. Особый интерес для нас представляют столбцы family_status, children, debt.

Для начала необходимо подготовить данные для анализа. Уже видно, что некоторые переменные содержат ошибки. Например, в информации о детях есть отрицательное значение и значение 20 (либо это очень многодетная семья, либо опечатка). Скорее всего в данных есть дубликаты. Уже сейчас мы знаем, что одинаковое образование записано в по-разному. 

В столбце с целью кредита много значений, но некоторые из них описывают одну и ту же цель одинаковыми словами. Такие повторы необходимо убрать и выделить схожие цели.

### Шаг 2. Предобработка данных

Исключим пропуски и проверим данные на наличие дубликатов.

### Обработка пропусков

In [6]:
print(df.isna().sum()) #проверяем количество пропусков вызовом набора методов для суммирования пропущенных значений.

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64


Пустые значения свидетельствуют, что у некоторых клиентов зафиксирована не вся информация. Отсутсвуют данные о зарплате и о трудовом стаже.  Причины могут быть разные: заемщик являлся безработным, его не спросили об этом или он скрыл эти данные намеренно. 

Люди часто не отвечают на вопрос о своих доходах. Образование сильно коррелирует с доходами, значит пропуски в поле доходы нельзя считать совершенно случайными. Выполняется гипотеза MAR. MAR (Missing At Random) — данные пропущены не случайно, а ввиду некоторых закономерностей.  В случае MAR исключение пропусков вполне приемлемо.

Однако, удаление почти 10% данных приведет к неправильным выводам. Пропуски можно заполнить пропуски можно средним арифметическим или медианой. Для дохода больше подойдет медиана, для стажа можно использовать среднее. 



При изучении переменных мы заметили отрицательные значения в столбце days_employed, заменим их на положительные значения с помощью метода abs()

In [7]:
df['days_employed'] = df['days_employed'].abs() #замена отрицательных чисел на абсолютное значение


<div style="border:solid PaleVioletRed 2px; padding: 20px">
Заменила отрицательные значения в столбце days_employed. Среднее значение действительно поменялосб
</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2 </h2>

Excellent

</div>

In [8]:
mean_days_employed = df['days_employed'].mean() #расчитаем среднее значение вызовом метода mean
print('Среднее значение трудового стажа, в днях: {}'.format(int(mean_days_employed)))

median_total_income = df['total_income'].median() #расчитаем медианное значение вызовом метода median
print('Медианное значение дохода: {}'.format(int(median_total_income)))

df['days_employed']= df['days_employed'].fillna(mean_days_employed) #заполним пропуски в столбце days_employed средним значением трудового стажа для нашей выборки
df['total_income']= df['total_income'].fillna(median_total_income) #пропуски в столбце total_income заменим медианным значением дохода
print(df.isna().sum()) #проверяем количество пропусков теперь


Среднее значение трудового стажа, в днях: 66914
Медианное значение дохода: 145017
children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64


### Вывод

Теперь наш набор данных не содержит пропусков. 

### Замена типа данных

Для удобства приведем все значения к целому числу - к типу int. Также заменим ошибочные отрицательные значения

In [9]:

df['days_employed'] = df['days_employed'].astype('int') #приведение к целому числу
print(df['days_employed'])
df['total_income'] = df['total_income'].astype('int') #приведение к целому числу
print(df['total_income'])

print(df['children'].unique()) #среди уникальных есть значение -1, можно предположить, что это опечатка
df['children'] = df['children'].abs() #замена отрицательных чисел на абсолютное значение
print(df['children'].unique())




0          8437
1          4024
2          5623
3          4124
4        340266
          ...  
21520      4529
21521    343937
21522      2113
21523      3112
21524      1984
Name: days_employed, Length: 21525, dtype: int64
0        253875
1        112080
2        145885
3        267628
4        158616
          ...  
21520    224791
21521    155999
21522     89672
21523    244093
21524     82047
Name: total_income, Length: 21525, dtype: int64
[ 1  0  3  2 -1  4 20  5]
[ 1  0  3  2  4 20  5]


<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Все правильно! Однако, исправление отрицательных значений я бы провел до замены пропусков, так, скорее всего, значение `mean_days_employed` поменяется

</div>

Наличие 20 детей - редкость. Можно допустить, что в один банк может обратиться несколько семей с ровно 20 детьми. Проверим, сколько строк содержат такое значение.


In [10]:
print(df['children'].value_counts()) 


0     14149
1      4865
2      2055
3       330
20       76
4        41
5         9
Name: children, dtype: int64


Одна семья с 20 детьми - еще возможно, несколько семей с 20 детьми - тоже вероятно, 76 семей - похоже на опечатку. Скорее, всего там лишний ноль. Заменим на число 2. 


In [11]:

df.loc[df['children']==20, 'children']=2 
print(df['children'].value_counts())

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64


### Вывод

Теперь в данных нет опечаток, которые могут исказить результат, а также используются более подходящие для вычислений типы данных.

### Обработка дубликатов

Сначала уберем уже найденные повторения в столбце education. Одинаковое образование написано в разных регистрах. 

In [12]:
print(df['education'].value_counts())

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
УЧЕНАЯ СТЕПЕНЬ             1
Ученая степень             1
Name: education, dtype: int64


Приведем все переменные к одному регистру.

In [13]:
df['education'] = df['education'].str.lower()
print(df['education'].value_counts())

среднее                15233
высшее                  5260
неоконченное высшее      744
начальное                282
ученая степень             6
Name: education, dtype: int64


Теперь проверим наличие дублирующих строк

In [14]:
print(df.duplicated().sum())  #проверяем кол-во повторяющихся строк
df = df.drop_duplicates().reset_index(drop=True)  #удаляем дубликаты с сохранением индексов
print(df.duplicated().sum()) #проверяем, что дубликатов не осталось

71
0


<div style="border:solid PaleVioletRed 2px; padding: 20px"> 
 Удаление дубликатов в столбце образования перенесла вверх. Количество дубликатов увеличилось почти на 20 строк!

<div class="alert alert-warning">
<h2> Комментарий ревьюера </h2>

Здесь я бы также приведение к нижнему регистру перенес выше, потому что это может повлиять на кол-во дубликатов

</div>

### Вывод

Дублирование значений в столбце образование может быть связано с тем, что данные вводят разные операторы, кто-то предпочитает писать все в верхнем регистре, а кто-то наоборот. Дубликаты строк могли возникнут из-за технического сбоя или задублированы при переносе данных в базу. 

### Лемматизация

Для анализа нам нужно выяснить, какие самые популярные цели для кредита. В сыром виде анализировать столбец purpose сложно: цели записаны произвольно. Необходимо привести их к одному виду. 

In [15]:
print(len(df['purpose'].unique())) #посчитаем сколько все уникальных записей целей кредита

38


In [16]:
purposes = df['purpose'].tolist()


Избавимся от одинаковых по смыслу дубликатов с помощью лемматизации, то есть приведем слова к словарной форме

In [17]:

purposes = ' '.join(purposes)

#импортируем библиотеку с функцией лемматизации на русском языке 
from pymystem3 import Mystem
m = Mystem()
#передаем функции в качестве аргумента - список целей 
lemmas = m.lemmatize(purposes)


#подчитаем число упоминаний лемматизируемых слов 
from collections import Counter
print(Counter(lemmas))



Counter({' ': 55023, 'недвижимость': 6351, 'покупка': 5897, 'жилье': 4460, 'автомобиль': 4306, 'образование': 4013, 'с': 2918, 'операция': 2604, 'свадьба': 2324, 'свой': 2230, 'на': 2222, 'строительство': 1878, 'высокий': 1374, 'получение': 1314, 'коммерческий': 1311, 'для': 1289, 'жилой': 1230, 'сделка': 941, 'дополнительный': 906, 'заниматься': 904, 'подержать': 853, 'проведение': 768, 'сыграть': 765, 'сдача': 651, 'семья': 638, 'собственный': 635, 'со': 627, 'ремонт': 607, 'приобретение': 461, 'профильный': 436, 'подержанный': 111, '\n': 1})


Добавим в датафрейм столбец, который будет содержать соответвующую столбцу purpose лемму

In [18]:
def lemmatize(col):
    return m.lemmatize(col)
                        
df['purpose_lemmas'] = df['purpose'].apply(lemmatize)                

print(df['purpose_lemmas'].head())

0                 [покупка,  , жилье, \n]
1       [приобретение,  , автомобиль, \n]
2                 [покупка,  , жилье, \n]
3    [дополнительный,  , образование, \n]
4               [сыграть,  , свадьба, \n]
Name: purpose_lemmas, dtype: object


<div style="border:solid PaleVioletRed 2px; padding: 20px">
Так логичнее, но о том, что мне понадобиться такой столбец я поняла только, когда дошла до вопроса про наличие зависимости. 
    </div>

### Вывод

Самая частая цель кредита - это приобретение недвижимости и жилья, на втором месте - покупка автомобиля, на третьем - образование. Затем идут операции и свадьбы.

### Категоризация данных

Наша основная задача - выяснить, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Выделим категории для анализа. 

Категории по наличию детей
- есть дети (1-2 ребенка)
- многодетные (3 и более)
- нет детей (0)

Категории по семейному положению:
- состоит в браке (отнесем сюда и гражданский, и зарегистрированный)
- не состоит в браке (отнесем сюда одиноких, вдовцов и в разводе)

In [19]:
#создадим функцию, которая будет определять категорию при выполнении условия

def children_category(kids):
    if kids == 0:
        return 'нет детей'
    if kids >= 1 and kids <= 2:
        return 'есть дети'
    if kids >= 3:
        return 'многодетные'
#добавим новый столбец с категориями children_category
df['children_category'] = df['children'].apply(children_category)

    

def marriage_status(family_status_id):
    if family_status_id in (0, 1):
        return 'в браке'
    if family_status_id in (2,3,4):
        return 'не в браке'
    #добавим новый столбец с категорией marriage_status

df['marriage_status'] = df['family_status_id'].apply(marriage_status)

print(df['marriage_status'].value_counts())
print(df['children_category'].value_counts())


в браке       16490
не в браке     4964
Name: marriage_status, dtype: int64
нет детей      14091
есть дети       6983
многодетные      380
Name: children_category, dtype: int64


Выделим категории по доходу. При разбиении будем ориентироваться на данные по зарплате в РФ. Средняя зарплата по стране - 43 400 рублей

In [20]:
print('max', df['total_income'].max())
print('min', df['total_income'].min())


def total_income_group(total_income):
    if total_income < 45000:
        return 'низкий доход'
    if total_income >= 45000 and total_income <= 80000:
        return 'средний доход'
    if total_income > 80000  and total_income <= 150000:
        return 'выше среднего'
    if total_income > 150000  and total_income <= 500000:
        return 'высокий'
    if total_income > 500000:
        return 'очень высокий'
#добавим новый столбец с категорией total_income_group
df['total_income_group'] = df['total_income'].apply(total_income_group)
df['total_income_group'].value_counts()

max 2265604
min 20667


выше среднего    9994
высокий          8962
средний доход    2059
очень высокий     222
низкий доход      217
Name: total_income_group, dtype: int64

Выделим категории по целям

In [21]:
def category_purpose(lemmas):
    for word in lemmas:
        if 'жилье' in word:
            return 'жилье'
        elif 'образование' in word:
            return 'образование'
        elif 'недвижимость' in word:
            return 'недвижимость'
        elif 'свадьба' in word:
            return 'свадьба'
        elif 'автомобиль' in word:
            return 'автомобиль'
        elif 'операция' in word:
            return 'операция'
        
    
    
                
df['category_purpose'] = df['purpose_lemmas'].apply(category_purpose)                

print(df['category_purpose'].head())

0          жилье
1     автомобиль
2          жилье
3    образование
4        свадьба
Name: category_purpose, dtype: object


<div class="alert alert-success">
<h2> Комментарий ревьюера </h2>

Категоризация выполнена на отлично

</div>

### Вывод

За кредитом обращалось больше людей, состоящих в браке, чем одиноких. С детьми кредит брали реже. Люди с низким доходом и с очень высоким доходом в выборке представлены меньше. Больше всего заемщиков с высоким и с доходом выше среднего. 

### Шаг 3. Ответьте на вопросы

- Есть ли зависимость между наличием детей и возвратом кредита в срок?

In [22]:
children_debt = df.pivot_table(index = ['children_category', 'debt'], values = 'children', aggfunc = 'count').reset_index()

#посчитаем долю каждой группы от общего количества пользователей
def share(col):
    return "{0:.2f}%".format(col/len(df)) 

children_debt['share']=children_debt['children'].apply(share) 
children_debt


    


children_category  debt  children  share
0         есть дети     0      6336  0.30%
1         есть дети     1       647  0.03%
2       многодетные     0       349  0.02%
3       многодетные     1        31  0.00%
4         нет детей     0     13028  0.61%
5         нет детей     1      1063  0.05%

<div style="border:solid PaleVioletRed 2px; padding: 20px"> 
Это попытка проделать группировку методом pivot_table, а не groupby. С методом все более менее понятно. В данной случае возник вопрос, какой столбец использовать для параметра values. Для функции count, кажется, подходил любой.
    
Как в таком случае негромоздко посчитать долю от группы, а не от всей выборки (следующий комментарий), я не знаю.
Придумалось только следующее очень объемное и путанное решение, но результат достигается. 
</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 </h2>

Решение здесь очень простое. Можно взять среднее по столбцу `debt` для каждой группы. Т.е так

</div>

In [41]:
df.pivot_table(index = 'children_category', 
               values = 'debt', 
               aggfunc = 'mean').reset_index().style.format({'debt': '{:.2%}'})

In [23]:
#выделим срезы из набора данных для разных категорий по наличию детей

children_debt_1 = children_debt[children_debt['children_category']=='есть дети']
children_debt_2 = children_debt[children_debt['children_category']=='нет детей']
children_debt_3 = children_debt[children_debt['children_category']=='многодетные']

#посчитаем численность групп

children_debt_1_sum = children_debt[children_debt['children_category']=='есть дети']['children'].sum()
children_debt_2_sum = children_debt[children_debt['children_category']=='нет детей']['children'].sum()
children_debt_3_sum = children_debt[children_debt['children_category']=='многодетные']['children'].sum()

#посчитаем относительные значения внутри группы.
#я не могу написать одну функцию для трех случаев, потому что не понимаю, 
#как передать аргумент children_debt_1_sum при методе apply

#получается ужасно, но получается :)

def share_in_group(col): 
    return "{0:.3f}%".format(col/children_debt_1_sum)
children_debt_1['share_in_group']=children_debt_1['children'].apply(share_in_group)


def share_in_group(col):
    return "{0:.3f}%".format(col/children_debt_2_sum)
children_debt_2['share_in_group']=children_debt_2['children'].apply(share_in_group)


def share_in_group(col):
    return "{0:.3f}%".format(col/children_debt_3_sum)
children_debt_3['share_in_group']=children_debt_3['children'].apply(share_in_group)


total = [children_debt_1, children_debt_2, children_debt_3 ]
total = pd.concat(total)
total

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:21: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:26: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:31: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/

children_category  debt  children  share share_in_group
0         есть дети     0      6336  0.30%         0.907%
1         есть дети     1       647  0.03%         0.093%
4         нет детей     0     13028  0.61%         0.925%
5         нет детей     1      1063  0.05%         0.075%
2       многодетные     0       349  0.02%         0.918%
3       многодетные     1        31  0.00%         0.082%

<div class="alert alert-warning">
<h2> Комментарий ревьюера v2 </h2>

При выполнении ячейки выше выходит ошибка `SettingWithCopyWarning: A value is trying to be set on a copy of a slice from a DataFrame`, подсвеченная розовым цветом (а значит, она важна для качества кода). Дело в том, что при применении `apply` используется цепное индексирование (т.е функция выполняется для каждого индекса), а в твоем случае, датафреймы имеют непоследовательную индексацию. Решить проблему позволит:
- сбрасывание индексов датафрейма, можно использовать [reset_index](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.reset_index.html)
- использование [loc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html) и [iloc](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html)
- можно отключить данное предупреждение прописав `pd.options.mode.chained_assignment = None` перед исполненем ячейки 😉

</div>

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Твоя реализация вывода результатов вполне рабочая, однако очень громоздкая. В данном случае более коротким решением будет построение [сводных таблиц](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html). Это очень важный и популярный, нужный инструмент, мне нужно убедиться, что ты умеешь им пользоваться. Реализуй то же самое используя `pivot_table()` и дай комментарии, насколько удобен для тебя инструмент.
    
</div>

### Вывод

Если выделять доли внутри группы, то получается, что доля задолжников с детьми выше, чем без.


Старый комментарий. Доля задолжников без детей выше, чем с детьми. Интересно посмотреть, как распределены доходы между заемщиками с детьми и без.


In [24]:
children_income = df.pivot_table(index = ['children_category', 'total_income_group'], values = 'children', aggfunc = ['count'])
children_income


count
                                     children
children_category total_income_group         
есть дети         высокий                2971
                  выше среднего          3297
                  низкий доход             58
                  очень высокий            85
                  средний доход           572
многодетные       высокий                 172
                  выше среднего           172
                  низкий доход              1
                  очень высокий             6
                  средний доход            29
нет детей         высокий                5819
                  выше среднего          6525
                  низкий доход            158
                  очень высокий           131
                  средний доход          1458

У клиентов с детьми в среднем высокий доход встречается чаще. 

- Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [25]:
marriage_debt = df.pivot_table(index = ['marriage_status', 'debt'], values = 'family_status', aggfunc = 'count').reset_index()

def share(col):
    return "{0:.2f}%".format(col/len(df)) 

marriage_debt['share']=marriage_debt['family_status'].apply(share)


marriage_debt_true = marriage_debt[marriage_debt['marriage_status']=='в браке']
marriage_debt_true_sum = marriage_debt[marriage_debt['marriage_status']=='в браке']['family_status'].sum()

def share_in_group(col):
    return "{0:.2f}%".format(col/marriage_debt_true_sum)

marriage_debt_true['share_in_group']=marriage_debt_true['family_status'].apply(share_in_group)

marriage_debt_false = marriage_debt[marriage_debt['marriage_status']=='не в браке']
marriage_debt_false_sum = marriage_debt[marriage_debt['marriage_status']=='не в браке']['family_status'].sum()

def share_in_group(col):
    return "{0:.2f}%".format(col/marriage_debt_false_sum)

marriage_debt_false['share_in_group']=marriage_debt_false['family_status'].apply(share_in_group)
total = [marriage_debt_false, marriage_debt_true]
total = pd.concat(total)
total

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  from ipykernel import kernelapp as app
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


marriage_status  debt  family_status  share share_in_group
2      не в браке     0           4542  0.21%          0.91%
3      не в браке     1            422  0.02%          0.09%
0         в браке     0          15171  0.71%          0.92%
1         в браке     1           1319  0.06%          0.08%

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

Также здесь ты смотришь относительно общего кол-ва пользователей кредита. Более информативным будет посчитать относительные значения внутри группы. Т.е найти сколько пользователей из группы `в браке` имеют задолженность (1319 / 16507)
    
</div>

<div style="border:solid PaleVioletRed 2px; padding: 20px"> 
 Мне кажется, что доля задолжников от общего количества заемщиков показательнее, потому что в любой группе доля людей без долга всегда выше.
Выше я корявым способом посчитала доли от категории, по ним я не могу сделать вывод. Возможно, лучше сравнивать долю должников конкретной категории от общего количества должников. Но кажется, что тут я конкретно запуталась. 
</div>


In [26]:
marriage_debt = df.pivot_table(index = ['marriage_status', 'debt'], values = 'family_status', aggfunc = 'count').reset_index()

def share(col):
    return "{0:.2f}%".format(col/len(df)) 

marriage_debt['share']=marriage_debt['family_status'].apply(share)


marriage_debt_true = marriage_debt[marriage_debt['debt']==1]
marriage_debt_true_sum = marriage_debt[marriage_debt['debt']==1]['family_status'].sum()

def share_in_group(col):
    return "{0:.2f}%".format(col/marriage_debt_true_sum)

marriage_debt_true['share_in_group']=marriage_debt_true['family_status'].apply(share_in_group)

marriage_debt_false = marriage_debt[marriage_debt['debt']==0]
marriage_debt_false_sum = marriage_debt[marriage_debt['debt']==0]['family_status'].sum()

def share_in_group(col):
    return "{0:.2f}%".format(col/marriage_debt_false_sum)

marriage_debt_false['share_in_group']=marriage_debt_false['family_status'].apply(share_in_group)
total = [marriage_debt_false, marriage_debt_true]
total = pd.concat(total)
total

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  from ipykernel import kernelapp as app
/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:23: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


marriage_status  debt  family_status  share share_in_group
0         в браке     0          15171  0.71%          0.77%
2      не в браке     0           4542  0.21%          0.23%
1         в браке     1           1319  0.06%          0.76%
3      не в браке     1            422  0.02%          0.24%

<div style="border:solid PaleVioletRed 2px; padding: 20px">  В столбце share_in_group видно, что доля людей в браке среди невозвращающих выше.  
</div>

Проблем с возвратом больше у состоящих в браке, чем у одиноких людей. Дополнительно посмотрим, как много задолжностей у женатых/замужных с детьми и без.

In [27]:
marriage_children = df.pivot_table(index = ['marriage_status', 'children_category', 'debt'], values = ['children'], aggfunc = ['count'])
display(marriage_children)


count
                                       children
marriage_status children_category debt         
в браке         есть дети         0        5395
                                  1         546
                многодетные       0         323
                                  1          28
                нет детей         0        9453
                                  1         745
не в браке      есть дети         0         941
                                  1         101
                многодетные       0          26
                                  1           3
                нет детей         0        3575
                                  1         318

### Вывод

Число задолжников без детей выше и среди женатых, и среди одиноких. 

- Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [28]:
income_debt = df.pivot_table(index = ['total_income_group', 'debt'], values = 'total_income', aggfunc = 'count').reset_index()

def share(col):
    return "{0:.2f}%".format(col/len(df)) 

income_debt['share']=income_debt['total_income'].apply(share)

income_debt

total_income_group  debt  total_income  share
0            высокий     0          8250  0.38%
1            высокий     1           712  0.03%
2      выше среднего     0          9153  0.43%
3      выше среднего     1           841  0.04%
4       низкий доход     0           203  0.01%
5       низкий доход     1            14  0.00%
6      очень высокий     0           208  0.01%
7      очень высокий     1            14  0.00%
8      средний доход     0          1899  0.09%
9      средний доход     1           160  0.01%

### Вывод

Люди с высоким доходом и с доходом выше среднего чаще имеют задолженность по кредиту, чем с низким, средним и очень высоким. Можно предположить, что те, кто получают мало, не готовы дополнительно платить и за просроченный платеж по кредиту, а у людей с высоким доходом проблем с выплатами не возникает.

- Как разные цели кредита влияют на его возврат в срок?

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

При выполнении работы важно соблюдать структуру, которая дана в описании проекта, она поможет тебе построить последовательное исследование. Поэтому перенеси, пожалуйста, лемматизацию в раздел лемматизации или категоризации 🙃.

</div>

In [29]:
purpose_debt = df.pivot_table(index = ['category_purpose', 'debt'], values = 'purpose', aggfunc = 'count').reset_index()

def share(col):
    return "{0:.2f}%".format(col/len(df)) 

purpose_debt['share']=purpose_debt['purpose'].apply(share)    
purpose_debt

category_purpose  debt  purpose  share
0        автомобиль     0     3903  0.18%
1        автомобиль     1      403  0.02%
2             жилье     0     3548  0.17%
3             жилье     1      260  0.01%
4      недвижимость     0     4082  0.19%
5      недвижимость     1      317  0.01%
6       образование     0     3643  0.17%
7       образование     1      370  0.02%
8          операция     0     2399  0.11%
9          операция     1      205  0.01%
10          свадьба     0     2138  0.10%
11          свадьба     1      186  0.01%

### Вывод



Самая высокая доля задолжников - те, кто брали кредит для покупки автомобиля.


### Шаг 4. Общий вывод

Перед началом анализа была проведена предобработка данных: 
- пропущенные значения в столбце days_employed были заменены на среднее значение и в столбце total_income на медианное значение. - удалены дубликаты в столбце education
- удалены повторяющиеся строки
- устраненны ошибочные отрицательные значения и опечатка в столбце children.

Также были выделенные ключевые слова для столбца purpose и опредлены самые популярные цели для кредитов - приобретение недвижимости и жилья,покупка автомобиля, образование, операции и свадьбы.

Данные были разделены на несколько категорий - по наличию детей, семейному положению, доходу и целям. В датафрейм добавлены соответствующие столбцы -children_category, marriage_status, total_income_group, category_purpose

Наличие детей не влияет на факт погашения кредита в срок, а вот их отстутвие, наоборот, повышает вероятность стать задолжником.
Задолженность по кредиту чаще возникает у людей женатых/замужних. 
У людей с высоким доходом и доходом выше среднего задолженности возникают чаще, чем у людей с низким, средним и очень высоким.
Цель кредита - автомобиль - самая популярная среди должников. 

<div class="alert alert-danger">
<h2> Комментарий ревьюера </h2>

В общий вывод стоит добавить результаты работ на втором и третьем шагах, как были обрабатаны пропуски, их возможные причины и тд. Это позволит людям, у которых нет времени вникать в работу, понять что было сделано и как.
    
</div>

<div style="border:solid purple 2px; padding: 20px">

У тебя получилась очень хорошая работа! Многие моменты сделаны просто прекрасно. Ты показываешь отличное владение изучаемым материалом: уверенно пользуешься pandas, умеешь подготавливать, очищать, обогощать данные. На протяжении всей работы чувствуется глубина проработки задачи и это безусловный плюс для нас.

Остались небольшие доработки:

- перенести лемматизацию в соответствующий раздел;
- В шаге 3 необходимо посчитать относительные групповые показатели;
- В шаге 3 необходимо получить результаты с использованием метода `pivot_table()`;
- В шаге 4 необходимо дополнить общий вывод.

Жду твоих исправлений, у тебя все получится 😊 

</div>

<div style="border:solid PaleVioletRed 2px; padding: 20px"> 
Лемматизацию и категоризацию по цели перенесла в соотствующие разделы, где им и место.
Использовала метод pivot_table для вывода результатов.
Общий вывод добавила.
Возникли трудности с подчетом относительныхгрупповых показателей. Посчитала, как смогла. Как сделать иначе, найти не удалось. 
Посчитала своим корявым методом для детей и семейного положения. Очень надеюсь на совет, как это сделать меньшим количесвтом строк.
И у меня вопрос - почему лучше считать относительные групповые показатели, а не, например, показатели долей категорий среди должников. 
    
Буду ждать ответа, спасибо!

</div>

<div class="alert alert-success">
<h2> Комментарий ревьюера v2 </h2>

По подсчету относительных показателей показал, как можно было решить короче. Надеюсь все понятно))) Что касается последнего вопроса, дело в том, что при подсчете относительно всего датафрейма не совсем понятно, какая категория, например, целей кредита лучше. Т.е банку нужно понять, на что выгоднее выдавать кредит, на автомобиль или на свадьбу. А это можно понять только посчитав относительные групповые показатели

</div>

### Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  открыт файл;
- [x]  файл изучен;
- [x]  определены пропущенные значения;
- [x]  заполнены пропущенные значения;
- [x] есть пояснение, какие пропущенные значения обнаружены;
- [x]  описаны возможные причины появления пропусков в данных;
- [x]  объяснено, по какому принципу заполнены пропуски;
- [x]  заменен вещественный тип данных на целочисленный;
- [x]  есть пояснение, какой метод используется для изменения типа данных и почему;
- [x]  удалены дубликаты;
- [x]  есть пояснение, какой метод используется для поиска и удаления дубликатов;
- [x] описаны возможные причины появления дубликатов в данных;
- [x]  выделены леммы в значениях столбца с целями получения кредита;
- [x]  описан процесс лемматизации;
- [x]  данные категоризированы;
- [x]  есть объяснение принципа категоризации данных;
- [x]  есть ответ на вопрос: "Есть ли зависимость между наличием детей и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между семейным положением и возвратом кредита в срок?";
- [x]  есть ответ на вопрос: "Есть ли зависимость между уровнем дохода и возвратом кредита в срок?";
- [x] есть ответ на вопрос: "Как разные цели кредита влияют на его возврат в срок?";
- [x]  в каждом этапе есть выводы;
- [x]  есть общий вывод.